README
NAME
    OpusVL::Preferences - Generic DBIC preferences module

VERSION
    Version 0.13

SYNOPSIS
    This is a really simple module to pull into result classes so you can
    attach preferences, rather than have to continually extend the schema
    definition where its probably not appropriate.

    Say you had an Employees class, and wanted to define the following
    preferences for a customer:

    grows_plants
    has_untidy_desk
    likes_noodles

    You would set up your Result class as follows:

            package Result::Employee;

            use strict;
            use Moose;
        
            extends 'DBIx::Class::Core';

            with 'OpusVL::Preferences::RolesFor::Result::PrfOwner';

            __PACKAGE__->prf_owner_init;

            ...

    And the ResultSet class would be:

            package ResultSet::Employee;

            use strict;
            use Moose;

            extends 'DBIx::Class::ResultSet';

            with 'OpusVL::Preferences::RolesFor::ResultSet::PrfOwner';

            ...

    This would initialise the class with 3 preferences, set to the
    appropriate defaults. Within the Employee class, the following methods
    are exposed to manage the preferences:

  Result Class Methods
   prf_get
    Get the current value of the preference (either the default or local
    copy as appropriate).

            $p = $employee->prf_get ('grows_plants');    # $p == 1

   prf_set
    Overides the default preference value for the employee in question:

            $employee = prf_set (grows_plants => 0);
            $p = $employee->prf_get ('grows_plants');    # $p == 0

   prf_reset
    Deletes any local overrides and uses the default

            $employee->prf_reset ('grows_plants');
            $p = $employee->prf_get ('grows_plants');    # $p == 1

   prf_preferences
    Returns a resultset containing PrfPreference classes.

  ResultSet Methods
   prf_defaults
    Returns a resultset of the default preferences setup for this resultset.
    Add more results to this object to add more defaults. For example, the
    following might be in the initdb routine:

            sub initdb
            {
                    my $self = shift;

                    $self->prf_defaults->populate
                    ([
                            { name => 'grown_plants'    => default_value => '1' },
                            { name => 'has_untidy_desk' => default_value => '1' },
                            { name => 'likes_noodles'   => default_value => '1' },
                    ]);
            }

   prf_search
    To be completed. Will allow an Employee resultset to be return using
    preferences as a search parameter.

AUTHOR
    Rich Doughty, "<rich at opusvl.com>"

BUGS
    None. Past, present and future.

SUPPORT
    You can find documentation for this module with the perldoc command.

        perldoc OpusVL::Preferences

ACKNOWLEDGEMENTS
LICENSE AND COPYRIGHT

Copyright 2011 Opus Vision Limited

This library is free software; you can redistribute it and/or modify it under the same terms as Perl itself.

If you require assistance, support, or further development of this software, please contact OpusVL using the details below:

Telephone: +44 (0)1788 298 410
Email: community@opusvl.com
Web: <http://opusvl.com



Changes
Revision history for OpusVL-Preferences

{{ $NEXT }}

0.33      2018-08-10 11:32:36+01:00 Europe/London

  * Created deployment for nullable fields

0.32      2018-08-10 10:32:36+01:00 Europe/London

  * Fixed display mask and mask char. Both fields are nullable. Mask char is
    only required if display mask is set.
  * Display mask and mask char no longer get set to zero if not provided.

0.31      2018-07-18 14:21:39+01:00 Europe/London

  * Boolean fields have blank first option in search mode
  * Email search no longer requires an email address
  * Generated forms now use bootstrap

0.30      2018-06-19 12:51:23+01:00 Europe/London

  * DEPRECATED. The changes below (and above) are for compatibility ONLY
  * Converted to Dist::Zilla
  * Now making use of FB11 brain architecture
  * Now is a standalone component
  * Returns its own Result objects, which contain the original object
  * Added "GDPR erasure" stuff

0.29      2018-05-09 15:35:17+01:00 Europe/London

  * Fix curiously deleted lines that made it not work

0.28      2018-05-09 14:35:46+01:00 Europe/London

  * [T13511] GDPR defautls added

0.27      2017-09-12 14:40:12+01:00 Europe/London

  * Fields are no longer masked if the mask happens to be the empty string

0.26      2016-12-14 14:19:22+00:00 Europe/London

  * Added missing pod

0.25      2016-12-14 13:35:17+00:00 Europe/London

  * Updated methods for pulling all the prefs out in datastructures to do 
    decryption and allow for a hash
  * Added Jenkinsfile for testing

0.24      2016-11-04 17:10:57+00:00 Europe/London

  * Updated dependency on Postgres module

0.23      2016-11-04 12:01:59+00:00 Europe/London

  * Added field mask function
  * Added new searchable flag to indicate that it makes sense to search 
    on a field.
  * Pod tweaks
  * Added helper methods for getting an encryption/decryption sub
  * Updated behaviour to throwing an error when the field prf_get tries to 
    get doesn't exist
  * Added support for basic symmetric crypto on values.

0.22      2016-10-26 10:08:43+01:00 Europe/London

  * Provide a clearer error message when a field wasn't setup.

0.21      2016-02-26 08:53:20+00:00 Europe/London

  * Fixed RT 112429, test failures on non english locales.
  * Switched bug tracking over to github.
  * Fixed some more pod.

0.20      2016-02-25 16:52:19+00:00 Europe/London

  * Tweaked license/support sections in the POD to 
    make them generate correctly with dzil.

0.19      2016-02-25 16:27:23+00:00 Europe/London

  * Fixed test broken when converting to Dist::Zilla.

0.18      2016-02-25 16:09:28+00:00 Europe/London

  * Converted to Dist::Zilla and prepared for public release.

0.01    2012-11-29

  * First version, released on an unsuspecting world.



lib/OpusVL/Preferences.pm
package OpusVL::Preferences;

use warnings;
use strict;

# ABSTRACT: Generic DBIC preferences module

our $VERSION = '0.34';


=head1 SYNOPSIS

This is a really simple module to pull into result classes so you can attach
preferences, rather than have to continually extend the schema definition where
its probably not appropriate.

Say you had an Employees class, and wanted to define the following preferences
for a customer:

=over

=item grows_plants

=item has_untidy_desk

=item likes_noodles

=back

You would set up your Result class as follows:

	package Result::Employee;

	use strict;
	use Moose;
	
	extends 'DBIx::Class::Core';

	with 'OpusVL::Preferences::RolesFor::Result::PrfOwner';

	__PACKAGE__->prf_owner_init;

	...

And the ResultSet class would be:
	
	package ResultSet::Employee;

	use strict;
	use Moose;

	extends 'DBIx::Class::ResultSet';

	with 'OpusVL::Preferences::RolesFor::ResultSet::PrfOwner';

	...

This would initialise the class with 3 preferences, set to the appropriate
defaults. Within the Employee class, the following methods are exposed to
manage the preferences:

=head2 Result Class Methods

=head3 prf_get

Get the current value of the preference (either the default or local copy as
appropriate).

	$p = $employee->prf_get ('grows_plants');    # $p == 1

=head3 prf_set

Overides the default preference value for the employee in question:

	$employee = prf_set (grows_plants => 0);
	$p = $employee->prf_get ('grows_plants');    # $p == 0

=head3 prf_reset

Deletes any local overrides and uses the default

	$employee->prf_reset ('grows_plants');
	$p = $employee->prf_get ('grows_plants');    # $p == 1

=head3 prf_preferences

Returns a resultset containing PrfPreference classes.

=head2 ResultSet Methods

=head3 prf_defaults

Returns a resultset of the default preferences setup for this resultset. Add
more results to this object to add more defaults. For example, the following
might be in the initdb routine:

	sub initdb
	{
		my $self = shift;

		$self->prf_defaults->populate
		([
			{ name => 'grown_plants'    => default_value => '1' },
			{ name => 'has_untidy_desk' => default_value => '1' },
			{ name => 'likes_noodles'   => default_value => '1' },
		]);
	}

=head3 prf_search

To be completed. Will allow an Employee resultset to be return using
preferences as a search parameter.

=head1 BUGS

None. Past, present and future.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc OpusVL::Preferences

If you require assistance, support, or further development of this software, please contact OpusVL using the details below:

=over 4

=item *

Telephone: +44 (0)1788 298 410

=item *

Email: community@opusvl.com

=item *

Web: L<http://opusvl.com>

=back

=head1 ACKNOWLEDGEMENTS

=cut

1; # End of OpusVL::Preferences
